core/services/gas_updater.go

```diff
@@ -57,91 +70,292 @@ type gasUpdater struct {
 	}
+
+	var reqs []rpc.BatchElem
+	for i := lowestBlockToFetch; i <= highestBlockToFetch; i++ {
+		// NOTE: To save rpc calls, don't fetch blocks we already have in the history
+		if _, exists := blocks[i]; exists {
+			continue
+		}
+
+		req := rpc.BatchElem{
+			Method: "eth_getBlockByNumber",
+			Args:   []interface{}{toBlockNumStr(i), true},
+			Result: &Block{}, // FIXME: Can this be a *((*Block)(nil)) instead?
+		}
+		reqs = append(reqs, req)
+	}
+
+	gu.logger.Debugf("GasUpdater: fetching %v blocks (%v in local history)", len(reqs), len(blocks))
+	err := gu.store.EthClient.BatchCallContext(ctx, reqs)
+	if err != nil {
+		return errors.Wrap(err, "GasUpdater#fetchBlocks error fetching blocks with BatchCallContext")
+	}
+
+	for _, req := range reqs {
+		result, err := req.Result, req.Error
+		if err != nil {
+			gu.logger.Warnw("GasUpdater#fetchBlocks ignoring errored block", "err", err, "blockNum", head.Number)
+			continue
+		}
+
+		b, is := result.(*Block)
+		if !is {
+			return errors.Errorf("expected result to be a %T, got %T", &Block{}, result)
+		}
+		if b == nil {
+			return errors.New("invariant violation: got nil block")
+		}
+		if b.Hash == (common.Hash{}) {
+			gu.logger.Warnw("GasUpdater#fetchBlocks block was missing hash", "block", b, "blockNum", head.Number, "erroredBlockNum", b.Number)
+			continue
+		}
+
+		blocks[b.Number] = *b
+	}
+
+	newBlockHistory := make([]Block, 0)
+	for _, block := range blocks {
+		newBlockHistory = append(newBlockHistory, block)
+	}
+	sort.Slice(newBlockHistory, func(i, j int) bool {
+		return newBlockHistory[i].Number < newBlockHistory[j].Number
+	})
+
+	start := len(newBlockHistory) - int(gu.rollingBlockHistorySize)
+	if start < 0 {
+		gu.logger.Infow(fmt.Sprintf("GasUpdater: using fewer blocks than the specified history size: %v/%v", len(newBlockHistory), gu.rollingBlockHistorySize), "rollingBlockHistorySize", gu.rollingBlockHistorySize, "headNum", head.Number, "blocksAvailable", len(newBlockHistory))
+		start = 0
+	}
+
+	gu.rollingBlockHistory = newBlockHistory[start:]
+
+	return nil
 }
 
-func (gu *gasUpdater) percentileGasPrice() int64 {
+func (gu *gasUpdater) percentileGasPrice() (int64, error) {
 	gasPrices := make([]int64, 0)
 	for _, block := range gu.rollingBlockHistory {
-		for _, tx := range block.Transactions() {
+		for _, tx := range block.Transactions {
 			gasPrices = append(gasPrices, tx.GasPrice().Int64())
 		}
 	}
+	if len(gasPrices) == 0 {
+		return 0, errors.New("no transactions")
+	}
 	sort.Slice(gasPrices, func(i, j int) bool { return gasPrices[i] < gasPrices[j] })
 	idx := ((len(gasPrices) - 1) * gu.percentile) / 100
 	for i := 0; i <= 100; i += 5 {
 		jdx := ((len(gasPrices) - 1) * i) / 100
 		promGasUpdaterAllPercentiles.WithLabelValues(fmt.Sprintf("%v%%", i)).Set(float64(gasPrices[jdx]))
 	}
-	return gasPrices[idx]
+	return gasPrices[idx], nil
 }
 
 func (gu *gasUpdater) setPercentileGasPrice(gasPrice int64) error {
```

Comment:
If the new price > EthMaxGasPriceWei  wouldn't we want to set it to the maximum here and log an error? The way it looks right now is that if a fast spike came in which put us over our max, we'd simply never update the gas price and it'd be stuck at the pre-spike level

Reply:
Yes, I think you might be right.